### DISCUSSION TITLE
MatrixWriter only saves when visualizer is open

### DISCUSSION THREAD

**Discussion by jbscherrer** [bonsai-rx/bonsai#1546] [Q&A]
I'm recording an AnalogInput using a NI DAQ and would like to save the last n samples to disk when a triggering event occurs. I've written a demo script (attached below) that should save the data when triggered by keyboard input. However, on its own this script only occasionally actually saves the data as desired. Strangely, if I open a visualizer for the save_audio SelectMany node below, the script suddenly starts saving most of the time it is triggered, although it still misses triggering events that occur in rapid succession. Any idea what's going on?

![image](https://github.com/bonsai-rx/bonsai/assets/39060975/88043115-730a-4582-a969-910bdc7b0ca0)

[test_save.zip](https://github.com/bonsai-rx/bonsai/files/12500915/test_save.zip)

**Comment by bruno-f-cruz** [comment]
You are probably closing the sequence faster than you can log the file.  Move the Take after the MatrixWriter and if that doesnt fix it, add a small delay after MatrixWriter.

**Reply by jbscherrer** [reply]
`Take` after `MatrixWriter` doesn't improve the save rate. Adding a 2 second delay after the MatrixWriter improves the save rate, but still only a random subset of files get saved successfully, maybe 50%. Increasing the delay beyond 2 seconds doesn't help. Any other ideas as to what might be happening?

**Reply by bruno-f-cruz** [reply]
Out of curiosity, how are you calculating this failure rate?

**Reply by jbscherrer** [reply]
Just counting the number of times I press the trigger key and the number of files that get saved.

**Reply by bruno-f-cruz** [reply]
Can you do the same but counting the number of non-zero size matrices that arrive at `snippet_acq`?


Also, almost certainly unrelated, but each time you click a Key, you are creating a new observable without disposing of it. Consider 

![image](https://github.com/bonsai-rx/bonsai/assets/7049351/a10feefe-3d55-44ea-a258-8c3282db8611)

Over 

![image](https://github.com/bonsai-rx/bonsai/assets/7049351/c6dbc2a2-694c-423e-b311-a36dda99c469)

**Reply by jbscherrer** [reply]
As far as I can tell, a non-zero-size matrix always propagates correctly to `snippet_acq`. The dropoff is apparently in the saving step.

**Reply by jbscherrer** [reply]
This is also a behavior that I've observed across multiple machines and both solid state and magnetic disk drives.

**Reply by glopesdev** [reply]
This is just for sanity check, but can you replace MatrixWriter with VideoWriter and change the Buffered property to False?

If needed specify a FourCC of "DIB " for raw data. The data files themselves won't make sense, but I would be curious if you can get the correct number of files with this.

I will try to have a look at the workflow later.

**Reply by jbscherrer** [reply]
Just did this--the video files get saved correctly every time.

**Reply by jbscherrer** [reply]
Were there any further developments on this? We recently tried using MatrixWriter again for a different project and ran into the same issue, where only a subset of trials were getting saved.
